client/my-sites/domains/domain-management/transfer/transfer-to-other-site/confirmation-dialog.jsx

```diff
@@ -11,25 +11,80 @@ import { get } from 'lodash';
 /**
  * Internal Dependencies
  */
-import { getSite } from 'calypso/state/sites/selectors';
+import { currentUserHasFlag } from 'calypso/state/current-user/selectors';
+import { getSite, isCurrentPlanPaid } from 'calypso/state/sites/selectors';
 import { Dialog } from '@automattic/components';
+import { NON_PRIMARY_DOMAINS_TO_FREE_USERS } from 'calypso/state/current-user/constants';
 
 class TransferConfirmationDialog extends React.PureComponent {
 	static propTypes = {
+		isMapping: PropTypes.bool.isRequired,
 		isVisible: PropTypes.bool.isRequired,
 		targetSiteId: PropTypes.number.isRequired,
 		disableDialogButtons: PropTypes.bool.isRequired,
 		domainName: PropTypes.string.isRequired,
 		onConfirmTransfer: PropTypes.func.isRequired,
 		onClose: PropTypes.func.isRequired,
+		primaryWithPlansOnly: PropTypes.bool,
 	};
 
 	onConfirm = ( closeDialog ) => {
 		this.props.onConfirmTransfer( this.props.targetSite, closeDialog );
 	};
 
+	getMessage() {
+		const { domainName, isMapping, targetSite, translate } = this.props;
+		const targetSiteTitle = get( targetSite, 'title', translate( 'Site Title' ) );
+		if ( isMapping ) {
+			return translate(
+				'Do you want to transfer mapping of {{strong}}%(domainName)s{{/strong}} ' +
+					'to site {{strong}}%(targetSiteTitle)s{{/strong}}?',
+				{
+					args: { domainName, targetSiteTitle },
+					components: { strong: <strong /> },
+				}
+			);
+		}
+
+		return translate(
+			'Do you want to transfer {{strong}}%(domainName)s{{/strong}} ' +
+				'to site {{strong}}%(targetSiteTitle)s{{/strong}}?',
+			{
+				args: { domainName, targetSiteTitle },
+				components: { strong: <strong /> },
+			}
+		);
+	}
+
+	renderTargetSiteOnFreePlanWarnings() {
+		if ( this.props.primaryWithPlansOnly ) {
```

Comment:
Just to be clear -- I think we still have a missing edge case here. (But let's fix separately so we can deploy this PR!)

If the owner of the target site is exempt from PWPO, then the domain can still be set as primary.